ocaml/typing/solver.ml

```diff
@@ -658,16 +676,157 @@ module Solver_mono (C : Lattices_mono) = struct
+          let src = C.src dst f in
+          generalize_topology ~log src ~current_level ~generic_level v
+        in
+        List.iter do_gen u.vupper;
+        List.iter do_gen u.vlower
+      end
+
+  let generalize_v
+      : type a. log:_ -> a C.obj -> current_level:int ->
+        generic_level:int -> a var -> unit =
+    fun ~log dst ~current_level ~generic_level u ->
+      generalize_topology ~log dst ~current_level ~generic_level u;
+      update_level_v ~log dst generic_level u
+
+  type varcond = { b : 'a. 'a var -> bool; } [@@unboxed]
+
+  let rec copy_under_condition_v
+      : type a. copylog:_ -> varcond -> ?current_level:int ->
+        a C.obj -> a var -> a var =
+    fun ~copylog cond ?current_level obj v ->
+      if not (cond.b v) then v
+      else begin
+        match v.optcopy with
+        | Some v' -> v'
+        | None ->
+          let copy = fresh ~upper:v.upper ~lower:v.lower ~level:v.level obj in
+          set_optcopy ~changes:copylog obj v (Some copy);
+          let vupper = List.map
+            (fun (Amorphvar (u, f)) ->
+              let src = C.src obj f in
+              let ucopy = copy_under_condition_v ~copylog cond ?current_level src u in
+              (Amorphvar (ucopy, f)))
+            v.vupper;
+          in
+          let vlower = List.map
+            (fun (Amorphvar (u, f)) ->
+              let src = C.src obj f in
+              let ucopy = copy_under_condition_v ~copylog cond ?current_level src u in
+              (Amorphvar (ucopy, f)))
+            v.vlower
+          in
+          copy.vupper <- vupper;
+          copy.vlower <- vlower;
+          Option.iter (fun level -> update_level_v ~log:None obj level copy) current_level;
+          copy
+      end
+
+  (* generalize_structure has three cases:
+    (1) if bounds are tight, the var is moved to generic_level
+    (2) if bounds are fully open, do nothing
+    (3) if bounds are non-trivial (neither tight nor fully open) we make a copy, move the
+        original var to generic_level, and mark the two variables as equivalent by adding
+        constraint arrows via [add_vlower] and [add_vupper]
+  *)
+  let generalize_structure_v
+      : type a. log:_ -> a C.obj -> current_level:int ->
+        generic_level:int -> a var -> unit =
+    fun ~log dst ~current_level ~generic_level u ->
+      if C.le dst u.upper u.lower then begin
+        (* the bounds are tight *)
+        generalize_topology ~log dst ~current_level ~generic_level u;
+        update_level_v ~log dst generic_level u
+      end else if
+        C.le dst (C.max dst) u.upper &&
+        C.le dst u.lower (C.min dst) then
+        (* the bounds are fully open *)
+        update_level_v ~log dst current_level u
+      else begin
+        (* the bounds are non-trivial *)
+        let cond = { b = fun v -> v.level > current_level } in
+        let copylog = fresh_mode_copy_scope () in
+        let _ = copy_under_condition_v ~copylog cond ~current_level dst u in
```

Comment:
I think using the `copylog` like this is a little too clever and probably it would be better to do things more directly. I also think that currently this code will not apply the tight bounds optimization above to all the reachable nodes, which I think that it should.

I think the following algorithm should roughly work. It requires making a version of `generalize_topology` specialized to the `generalize_struct` case. To process a node it should:
1. Raise the level of the node to be above generic level
2. Recursively `generalize_topology_struct` all the generalizable variables reachable through `vlower` and `vupper`.(Note that at this point the reachable graph should be a valid graph, although some variables have been made polymorphic that we would rather were not).
3. Create a new node with the same level, lower and upper as the current node. Add the new node to the `vlower` of the current node and add the current node to the `vlower` of the new node. (Note that the graph is still valid).
4. Raise the level of the new node to current level using `update_level`.

Reply:
I missed out that `generalize_topology_struct` should do nothing if the node is already generic and it should not bother making the new node if the bounds are tight.